src/Haskell/Ide/Engine/Transport/LspStdio.hs

```diff
@@ -98,33 +107,49 @@ run
   -> IO Int
 run dispatcherProc cin _origDir plugins captureFp = flip E.catches handlers $ do
 
-  rin <- atomically newTChan :: IO (TChan ReactorInput)
+  rin        <- atomically newTChan :: IO (TChan ReactorInput)
   commandIds <- allLspCmdIds plugins
 
   let dp lf = do
         cancelTVar  <- atomically $ newTVar S.empty
         wipTVar     <- atomically $ newTVar S.empty
         versionTVar <- atomically $ newTVar Map.empty
+        diagIn      <- atomically newTChan
         let dEnv = DispatcherEnv
               { cancelReqsTVar = cancelTVar
               , wipReqsTVar    = wipTVar
               , docVersionTVar = versionTVar
               }
-        let reactorFunc = runReactor lf dEnv cin diagnosticProviders hps sps
-              $ reactor rin
+        let react = runReactor lf dEnv cin diagIn diagnosticProviders hps sps
```

Comment:
Are there any caveats to look out for when running a second instance of R ?

Reply:
Good point.  In principle it need not be in R, if we encapsulate things properly

Reply:
runReactor is almost an alias for runReaderT, so only one reactor is actually running at a time.

Reply:
And the TVar updates are protected by `atomically`, so I guess at first blush it should be ok.

The danger comes if we ever change the definition of `R`, and forget/do not realise that it is used twice.

Reply:
@bubba what do you think?

Reply:
Is it possible to just pass the diagnostics chan to `queueDiagnosticsRequest`/`Debounce.forMonoid` without the rest of the stuff in `R`?

Reply:
I can try removing the new chan from `R` if you both agree it is a good idea.

Reply:
I think not running two instances of a monad with a shared state is a good idea.

I *do* think that in this case it is safe, given that it is a `runReaderT`, and we are using `TVar`s which are accessed atomically.  But that is only the case now, things may change in future.

Reply:
And my words are not all that clear. I am with @bubba on this one.

Reply:
Cool, I'll change it then